
# Battle of the Neighbourhoods - Toronto

In [1]:
import pandas as pd

# Question 1

## Importing Data

In [2]:
import requests

In [3]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
wiki_url = requests.get(url)
wiki_url

<Response [200]>

Response 200 means that we are able to make the connection to the page

In [4]:
wiki_data = pd.read_html(wiki_url.text)
wiki_data

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                          Neighbourhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

In [5]:
len(wiki_data), type(wiki_data)

(3, list)

We need the first table alone, so dropping the other tables

In [6]:
wiki_data = wiki_data[0]
wiki_data

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

Dropping Borough which are not assigned

In [7]:
df = wiki_data[wiki_data["Borough"] != "Not assigned"]
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

Grouping the records based on Postal Code

In [8]:
df = df.groupby(['Postal Code']).head()
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

Checking for number of records where Neighbourhood is "Not assigned"

In [9]:
df.Neighbourhood.str.count("Not assigned").sum()

0

In [10]:
df = df.reset_index()
df

index Postal Code           Borough  \
0        2         M3A        North York   
1        3         M4A        North York   
2        4         M5A  Downtown Toronto   
3        5         M6A        North York   
4        6         M7A  Downtown Toronto   
..     ...         ...               ...   
98     160         M8X         Etobicoke   
99     165         M4Y  Downtown Toronto   
100    168         M7Y      East Toronto   
101    169         M8Y         Etobicoke   
102    178         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 4 columns]

In [11]:
df.drop(['index'], axis = 'columns', inplace = True)
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [12]:
df.shape

(103, 3)

Answer to Question 1: We have 103 rows and 3 columns

# Question 2

Installing geocoder

In [13]:
pip install geocoder

Note: you may need to restart the kernel to use updated packages.


In [14]:
import geocoder # import geocoder

Tried the below approach, ran for 20 mins, then killed it. Changing the code cell to Text for now so that the run all execution doesn't stop.

```python
# initialize your variable to None
lat_lng_coords = None

postal_code = 'M5G'

# loop until you get the coordinates
while(lat_lng_coords is None):
  g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
  lat_lng_coords = g.latlng

latitude = lat_lng_coords[0]
longitude = lat_lng_coords[1]
```

Alternatively, as suggested in the assignment, Importing the CSV file from the URL

In [15]:
data = pd.read_csv("https://cocl.us/Geospatial_data")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [16]:
print("The shape of our wiki data is: ", df.shape)
print("the shape of our csv data is: ", data.shape)

The shape of our wiki data is:  (103, 3)
the shape of our csv data is:  (103, 3)


Since the dimensions are the same, we can try to join on the postal codes to get the required data.

Checking the column types of both the dataframes, especially Postal Code column since we are trying to join on it

In [17]:
df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

In [18]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [19]:
combined_data = df.join(data.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [20]:
combined_data.shape

(103, 5)

**Solution:** We get 103 rows as expected when we do a inner join, so we have good data.

# Question 3

Drawing inspiration from the previous lab where we cluster the neighbourhood of NYC, We cluster Toronto based on the similarities of the venues categories using Kmeans clustering and Foursquare API.

In [21]:
import sys
!conda install --yes --prefix {sys.prefix} -c conda-forge geopy
!conda install --yes --prefix {sys.prefix} -c conda-forge folium
!conda install --yes --prefix {sys.prefix} -c conda-forge scikit-learn
from geopy.geocoders import Nominatim 

Solving environment: ...working... done

# All requested packages already installed.

Solving environment: ...working... done

# All requested packages already installed.

Solving environment: ...working... done

## Package Plan ##

  environment location: C:\Users\user\.conda\envs\test

  added / updated specs:
    - scikit-learn


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    joblib-1.0.0               |     pyhd8ed1ab_0         206 KB  conda-forge
    libblas-3.9.0              |1_h8933c1f_netlib         193 KB  conda-forge
    libcblas-3.9.0             |3_hd5c7e75_netlib          95 KB  conda-forge
    liblapack-3.9.0            |3_hd5c7e75_netlib         2.7 MB  conda-forge
    scikit-learn-0.24.0        |   py38ha09990b_0         6.5 MB  conda-forge
    scipy-1.6.0                |   py38h5f893b4_0        23.3 MB  conda-forge
    threadpoolctl-2.1.0        |     pyh5ca1d4c_0    

In [22]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


Let's visualize the map of Toronto

In [23]:
import folium

In [24]:
# Creating the map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

Initializing Foursquare API credentials

In [25]:
CLIENT_ID = '5FPKISQSMANMLNH4JPKQPDGWWSST5NREDQL5ETWK4PLA35I3' 
CLIENT_SECRET = 'GW2E4ZND5FRDSQFHVDSVTYGFWFXW521UVLCUSJS0F4OPSYBI'
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 5FPKISQSMANMLNH4JPKQPDGWWSST5NREDQL5ETWK4PLA35I3
CLIENT_SECRET:GW2E4ZND5FRDSQFHVDSVTYGFWFXW521UVLCUSJS0F4OPSYBI


Next, we create a function to get all the venue categories in Toronto

In [26]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

Collecting the venues in Toronto for each Neighbourhood

In [27]:
venues_in_toronto = getNearbyVenues(combined_data['Neighbourhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [28]:
venues_in_toronto.shape

(1320, 5)

So we have 1320 records and 5 columns. Checking sample data


In [29]:
venues_in_toronto.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2         Parkwoods               43.753259               -79.329656   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue         Venue Category  
0         Brookbanks Park                   Park  
1         Brookbanks Pool                   Pool  
2           Variety Store      Food & Drink Shop  
3  Victoria Village Arena           Hockey Arena  
4               Portugril  Portuguese Restaurant

Checking the Venues based on Neighbourhood

In [30]:
venues_in_toronto.groupby('Neighbourhood').head()

Neighbourhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                             Parkwoods   
3                                      Victoria Village   
4                                      Victoria Village   
...                                                 ...   
1305  Mimico NW, The Queensway West, South of Bloor,...   
1306  Mimico NW, The Queensway West, South of Bloor,...   
1307  Mimico NW, The Queensway West, South of Bloor,...   
1308  Mimico NW, The Queensway West, South of Bloor,...   
1309  Mimico NW, The Queensway West, South of Bloor,...   

      Neighbourhood Latitude  Neighbourhood Longitude                   Venue  \
0                  43.753259               -79.329656         Brookbanks Park   
1                  43.753259               -79.329656         Brookbanks Pool   
2                  43.753259               -79.329656           Variety Store   
3                  43.725882               -79.315572  Victoria Village Arena   
4                  43.725882               -79.315572               Portugril   
...                      ...                      ...                     ...   
1305               43.628841               -79.520999              Wingporium   
1306               43.628841               -79.520999        South St. Burger   
1307               43.628841               -79.520999               Dollarama   
1308               43.628841               -79.520999          Healthy Planet   
1309               43.628841               -79.520999    Artisano Bakery Café   

             Venue Category  
0                      Park  
1                      Pool  
2         Food & Drink Shop  
3              Hockey Arena  
4     Portuguese Restaurant  
...                     ...  
1305            Wings Joint  
1306           Burger Joint  
1307         Discount Store  
1308        Supplement Shop  
1309                 Bakery  

[412 rows x 5 columns]

So there are 412 records for each neighbourhood.

Checking for the maximum venue categories

In [31]:
venues_in_toronto.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                          
Accessories Store                    Lawrence Manor, Lawrence Heights   
Airport                                                     Downsview   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Gate        CN Tower, King and Spadina, Railway Lands, Har...   
Airport Lounge      CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Warehouse Store                                      Thorncliffe Park   
Wine Bar                     Toronto Dominion Centre, Design Exchange   
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...   
Women's Store                                     Caledonia-Fairbanks   
Yoga Studio                            University of Toronto, Harbord   

                    Neighbourhood Latitude  Neighbourhood Longitude  \
Venue Category                                                        
Accessories Store                43.718518               -79.464763   
Airport                          43.737473               -79.394420   
Airport Food Court               43.628947               -79.394420   
Airport Gate                     43.628947               -79.394420   
Airport Lounge                   43.628947               -79.394420   
...                                    ...                      ...   
Warehouse Store                  43.705369               -79.349372   
Wine Bar                         43.653206               -79.379817   
Wings Joint                      43.628841               -79.520999   
Women's Store                    43.689026               -79.453512   
Yoga Studio                      43.715383               -79.321558   

                                              Venue  
Venue Category                                       
Accessories Store               Ardene Shoes Outlet  
Airport             Toronto Downsview Airport (YZD)  
Airport Food Court                Billy Bishop Café  
Airport Gate                                 Gate 8  
Airport Lounge                        Porter Lounge  
...                                             ...  
Warehouse Store                              Costco  
Wine Bar                          The National Club  
Wings Joint                              Wingporium  
Women's Store                         Maximum Woman  
Yoga Studio                               YogaSpace  

[235 rows x 4 columns]

There are around 235 different types of Venue Categories. Interesting!

## One Hot encoding the venue Categories

In [32]:
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Accessories Store  Airport  Airport Food Court  Airport Gate  \
0                     0        0                   0             0   
1                     0        0                   0             0   
2                     0        0                   0             0   
3                     0        0                   0             0   
4                     0        0                   0             0   
...                 ...      ...                 ...           ...   
1315                  0        0                   0             0   
1316                  0        0                   0             0   
1317                  0        0                   0             0   
1318                  0        0                   0             0   
1319                  0        0                   0             0   

      Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0                  0                0                 0                    0   
1                  0                0                 0                    0   
2                  0                0                 0                    0   
3                  0                0                 0                    0   
4                  0                0                 0                    0   
...              ...              ...               ...                  ...   
1315               0                0                 0                    0   
1316               0                0                 0                    0   
1317               0                0                 0                    0   
1318               0                0                 0                    0   
1319               0                0                 0                    0   

      Antique Shop  Aquarium  ...  Truck Stop  Vegetarian / Vegan Restaurant  \
0                0         0  ...           0                              0   
1                0         0  ...           0                              0   
2                0         0  ...           0                              0   
3                0         0  ...           0                              0   
4                0         0  ...           0                              0   
...            ...       ...  ...         ...                            ...   
1315             0         0  ...           0                              0   
1316             0         0  ...           0                              0   
1317             0         0  ...           0                              0   
1318             0         0  ...           0                              0   
1319             0         0  ...           0                              0   

      Video Game Store  Video Store  Vietnamese Restaurant  Warehouse Store  \
0                    0            0                      0                0   
1                    0            0                      0                0   
2                    0            0                      0                0   
3                    0            0                      0                0   
4                    0            0                      0                0   
...                ...          ...                    ...              ...   
1315                 0            0                      0                0   
1316                 0            0                      0                0   
1317                 0            0                      0                0   
1318                 0            0                      0                0   
1319                 0            0                      0                0   

      Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0            0            0              0            0  
1            0            0              0            0  
2            0            0              0            0  
3            0            0              0            0  
4   

Adding the neighbourhood to the encoded dataframe

In [33]:
toronto_venue_cat['Neighbourhood'] = venues_in_toronto['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Neighbourhood  Accessories Store  Airport  Airport Food Court  \
0         Parkwoods                  0        0                   0   
1         Parkwoods                  0        0                   0   
2         Parkwoods                  0        0                   0   
3  Victoria Village                  0        0                   0   
4  Victoria Village                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  ...  Truck Stop  \
0                    0             0  ...           0   
1                    0             0  ...           0   
2                    0             0  ...           0   
3                    0             0  ...           0   
4                    0             0  ...           0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                              0                 0            0   
1                              0                 0            0   
2                              0                 0            0   
3                              0                 0            0   
4                              0                 0            0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                      0                0         0            0   
1                      0                0         0            0   
2                      0                0         0            0   
3                      0                0         0            0   
4                      0                0         0            0   

   Women's Store  Yoga Studio  
0              0            0  
1              0            0  
2              0            0  
3              0            0  
4              0            0  

[5 rows x 236 columns]

We will group the Neighbourhoods, calculate the mean venue categories in each Neighbourhood 

In [34]:
toronto_grouped = toronto_venue_cat.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0      0.0                 0.0           0.0             0.0              0.0   
1      0.0                 0.0           0.0             0.0              0.0   
2      0.0                 0.0           0.0             0.0              0.0   
3      0.0                 0.0           0.0             0.0              0.0   
4      0.0                 0.0           0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Antique Shop  ...  Truck Stop  \
0               0.0             0.000000           0.0  ...         0.0   
1               0.0             0.000000           0.0  ...         0.0   
2               0.0             0.000000           0.0  ...         0.0   
3               0.0             0.000000           0.0  ...         0.0   
4               0.0             0.047619           0.0  ...         0.0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                            0.0               0.0          0.0   
1                            0.0               0.0          0.0   
2                            0.0               0.0          0.0   
3                            0.0               0.0          0.0   
4                            0.0               0.0          0.0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                    0.0              0.0       0.0          0.0   
1                    0.0              0.0       0.0          0.0   
2                    0.0              0.0       0.0          0.0   
3                    0.0              0.0       0.0          0.0   
4                    0.0              0.0       0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 236 columns]

Let's make a function to get the top most common venue categories

In [35]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [36]:
import numpy as np

There are way too many venue categories, we can take the top 10 to cluster the neighbourhoods

In [37]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                  Bank   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East           Coffee Shop   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0  Latin American Restaurant        Breakfast Spot          Skating Rink   
1                Coffee Shop        Sandwich Place                   Pub   
2                Coffee Shop              Pharmacy            Restaurant   
3                       Bank    Chinese Restaurant   Japanese Restaurant   
4             Sandwich Place    Italian Restaurant       Thai Restaurant   

  5th Most Common Venue      6th Most Common Venue 7th Most Common Venue  \
0           Yoga Studio       Ethiopian Restaurant           Escape Room   
1          Skating Rink                        Gym             Drugstore   
2           Supermarket  Middle Eastern Restaurant         Deli / Bodega   
3           Yoga Studio                Event Space  Ethiopian Restaurant   
4              Pharmacy                Pizza Place                   Pub   

  8th Most Common Venue        9th Most Common Venue  \
0     Electronics Store  Eastern European Restaurant   
1            Donut Shop                      Dog Run   
2        Ice Cream Shop                  Bridal Shop   
3           Escape Room            Electronics Store   
4            Restaurant                         Café   

        10th Most Common Venue  
0                    Drugstore  
1                Deli / Bodega  
2           Chinese Restaurant  
3  Eastern European Restaurant  
4                      Butcher

Let's make the model to cluster our Neighbourhoods

In [38]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [39]:
# set number of clusters
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(n_clusters=5, random_state=0)

Checking the labelling of our model

In [40]:
kmeans.labels_[0:100]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 0, 1, 1, 1, 4, 1, 1, 1, 1, 1, 1, 1, 1, 2, 4, 1, 1, 1,
       1, 1, 0, 1, 1, 1, 4, 1, 4, 1, 0, 1, 1, 1, 2, 1, 1, 0, 1, 1, 1, 4,
       3, 3, 1, 1, 4, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 0, 1, 1, 1, 1, 0])

Let's add the clustering Label column to the top 10 common venue categories

In [41]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Join toronto_grouped with combined_data on neighbourhood to add latitude & longitude for each neighborhood to prepare it for plotting

In [42]:
toronto_merged = combined_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656             0.0                  Pool   
1  43.725882 -79.315572             1.0          Intersection   
2  43.654260 -79.360636             1.0           Coffee Shop   
3  43.718518 -79.464763             1.0        Clothing Store   
4  43.662301 -79.389494             1.0           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
0                  Park     Food & Drink Shop            Yoga Studio   
1     French Restaurant           Coffee Shop  Portuguese Restaurant   
2                Bakery                  Park         Breakfast Spot   
3     Accessories Store           Coffee Shop               Boutique   
4      Sushi Restaurant        Sandwich Place     Italian Restaurant   

  5th Most Common Venue   6th Most Common Venue  7th Most Common Venue  \
0      Department Store             Escape Room      Electronics Store   
1          Hockey Arena     Distribution Center     Dim Sum Restaurant   
2                  Café             Yoga Studio      French Restaurant   
3           Event Space  Furniture / Home Store  Vietnamese Restaurant   
4   Japanese Restaurant                Beer Bar   Fast Food Restaurant   

         8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
0  Eastern European Restaurant              Drugstore             Donut Shop  
1                        Diner         Discount Store            Yoga Studio  
2                          Pub  Performing Arts Venue          Historic Site  
3           Dim Sum Restaurant             Comic Shop            Escape Room  
4                Smoothie Shop          Burrito Place    Distribution Center

Drop all the NaN values to prevent data skew

In [43]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

Plotting the clusters on the map

In [45]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [46]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

Let's verify each of our clusters

Cluster 1

In [47]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0        North York             0.0                  Pool   
21             York             0.0                  Park   
35        East York             0.0                  Park   
49       North York             0.0                  Park   
52       North York             0.0                   Gym   
61  Central Toronto             0.0                  Park   
66       North York             0.0                  Park   

         2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                         Park     Food & Drink Shop           Yoga Studio   
21               Women's Store                  Pool           Yoga Studio   
35           Convenience Store          Intersection           Yoga Studio   
49  Construction & Landscaping                Bakery           Yoga Studio   
52                        Park          Dessert Shop           Event Space   
61                 Swim School              Bus Line           Yoga Studio   
66           Convenience Store                   Bar           Yoga Studio   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0       Department Store           Escape Room     Electronics Store   
21      Department Store           Escape Room     Electronics Store   
35          Dessert Shop  Ethiopian Restaurant           Escape Room   
49    Dim Sum Restaurant           Event Space  Ethiopian Restaurant   
52  Ethiopian Restaurant           Escape Room     Electronics Store   
61          Dessert Shop           Escape Room     Electronics Store   
66    Dim Sum Restaurant           Event Space  Ethiopian Restaurant   

          8th Most Common Venue        9th Most Common Venue  \
0   Eastern European Restaurant                    Drugstore   
21  Eastern European Restaurant                    Drugstore   
35            Electronics Store  Eastern European Restaurant   
49                  Escape Room            Electronics Store   
52  Eastern European Restaurant                    Drugstore   
61  Eastern European Restaurant                    Drugstore   
66                  Escape Room            Electronics Store   

         10th Most Common Venue  
0                    Donut Shop  
21                   Donut Shop  
35                    Drugstore  
49  Eastern European Restaurant  
52                   Donut Shop  
61                   Donut Shop  
66  Eastern European Restaurant

Cluster 2

In [48]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1          North York             1.0          Intersection   
2    Downtown Toronto             1.0           Coffee Shop   
3          North York             1.0        Clothing Store   
4    Downtown Toronto             1.0           Coffee Shop   
6         Scarborough             1.0  Fast Food Restaurant   
..                ...             ...                   ...   
97   Downtown Toronto             1.0                  Café   
98          Etobicoke             1.0                  Pool   
99   Downtown Toronto             1.0           Coffee Shop   
100      East Toronto             1.0    Light Rail Station   
102         Etobicoke             1.0            Kids Store   

    2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
1       French Restaurant           Coffee Shop  Portuguese Restaurant   
2                  Bakery                  Park         Breakfast Spot   
3       Accessories Store           Coffee Shop               Boutique   
4        Sushi Restaurant        Sandwich Place     Italian Restaurant   
6            Dessert Shop           Event Space   Ethiopian Restaurant   
..                    ...                   ...                    ...   
97             Restaurant           Coffee Shop     Seafood Restaurant   
98                  River           Yoga Studio          Deli / Bodega   
99              Bookstore   Japanese Restaurant       Sushi Restaurant   
100           Yoga Studio            Skate Park          Auto Workshop   
102     Convenience Store         Tanning Salon                 Bakery   

      5th Most Common Venue   6th Most Common Venue  \
1              Hockey Arena     Distribution Center   
2                      Café             Yoga Studio   
3               Event Space  Furniture / Home Store   
4       Japanese Restaurant                Beer Bar   
6               Escape Room       Electronics Store   
..                      ...                     ...   
97   Gluten-free Restaurant               Speakeasy   
98              Escape Room       Electronics Store   
99        Indian Restaurant     Martial Arts School   
100                 Brewery           Burrito Place   
102         Supplement Shop          Discount Store   

           7th Most Common Venue 8th Most Common Venue  9th Most Common Venue  \
1             Dim Sum Restaurant                 Diner         Discount Store   
2              French Restaurant                   Pub  Performing Arts Venue   
3          Vietnamese Restaurant    Dim Sum Restaurant             Comic Shop   
4           Fast Food Restaurant         Smoothie Shop          Burrito Place   
6    Eastern European Restaurant             Drugstore             Donut Shop   
..                           ...                   ...                    ...   
97                    Steakhouse                   Pub                 Bakery   
98   Eastern European Restaurant             Drugstore             Donut Shop   
99                    Restaurant                  Café               Beer Bar   
100                   Comic Shop        Farmers Market   Fast Food Restaurant   
102       Thrift / Vintage Store        Sandwich Place           Burger Joint   

    10th Most Common Venue  
1              Yoga Studio  
2            Historic Site  
3              Escape Room  
4      Distribution Center  
6                  Dog Run  
..                     ...  
97              Food Court  
98                 Dog Run  
99    Ethiopian Restaurant  
100                 Garden  
102   Fast Food Restaurant  

[82 rows x 12 columns]

Cluster 3

In [49]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
57   North York             2.0        Baseball Field           Yoga Studio   
101   Etobicoke             2.0                  Pool        Baseball Field   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
57         Farmers Market           Event Space  Ethiopian Restaurant   
101           Yoga Studio          Dessert Shop  Ethiopian Restaurant   

    6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
57            Escape Room     Electronics Store  Eastern European Restaurant   
101           Escape Room     Electronics Store  Eastern European Restaurant   

    9th Most Common Venue 10th Most Common Venue  
57              Drugstore             Donut Shop  
101             Drugstore             Donut Shop

Cluster 4

In [50]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
12      Scarborough             3.0                   Bar   
62  Central Toronto             3.0          Home Service   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
12          Home Service           Yoga Studio    Dim Sum Restaurant   
62                Garden           Yoga Studio      Department Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
12           Event Space  Ethiopian Restaurant           Escape Room   
62  Ethiopian Restaurant           Escape Room     Electronics Store   

          8th Most Common Venue        9th Most Common Venue  \
12            Electronics Store  Eastern European Restaurant   
62  Eastern European Restaurant                    Drugstore   

   10th Most Common Venue  
12              Drugstore  
62             Donut Shop

Cluster 5

In [51]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
16              York             4.0            Playground   
32       Scarborough             4.0            Playground   
68   Central Toronto             4.0                  Park   
83   Central Toronto             4.0          Tennis Court   
85       Scarborough             4.0            Playground   
91  Downtown Toronto             4.0                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
16                 Field          Hockey Arena                 Trail   
32           Pizza Place    College Rec Center       College Stadium   
68      Sushi Restaurant         Jewelry Store                 Trail   
83            Playground                  Park                 Trail   
85                  Park          Intersection           Yoga Studio   
91            Playground                 Trail           Yoga Studio   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
16           Yoga Studio   Distribution Center    Dim Sum Restaurant   
32  Ethiopian Restaurant           Escape Room     Electronics Store   
68           Yoga Studio          Dessert Shop           Escape Room   
83        Discount Store      Department Store          Dessert Shop   
85      Department Store           Escape Room     Electronics Store   
91      Department Store           Escape Room     Electronics Store   

          8th Most Common Venue        9th Most Common Venue  \
16                        Diner               Discount Store   
32  Eastern European Restaurant                    Drugstore   
68            Electronics Store  Eastern European Restaurant   
83           Dim Sum Restaurant                        Diner   
85  Eastern European Restaurant                    Drugstore   
91  Eastern European Restaurant                    Drugstore   

   10th Most Common Venue  
16             Donut Shop  
32             Donut Shop  
68              Drugstore  
83            Yoga Studio  
85             Donut Shop  
91             Donut Shop

We have successfully cluster Toronto neighbourhood based on venue categories!